### PR TITLE
Changes to run px-backup tests on GKE cloud

### PR DESCRIPTION
**Pull request by sn-px**
**What this PR does / why we need it**:
Changes to run px-backup tests on GKE cloud

1. Made changes to connect to  gke cluster using kubeconfig
2. Fixed switching from source to destination cluster
3. Switching back to source from destination cluster 
4. Added bringing up of debug pods on destination cluster

**Which issue(s) this PR fixes** (optional)
Closes #https://portworx.atlassian.net/browse/PA-1307
https://portworx.atlassian.net/browse/PA-1659
https://portworx.atlassian.net/browse/PA-1670

Latest logs which covers switching between source and destination cluster : https://jenkins.pwx.dev.purestorage.com/job/Users/job/santoshn/job/gke-pipleine-san/421/consoleFull

Latest log : https://jenkins.pwx.dev.purestorage.com/job/Users/job/santoshn/job/gke-pipleine-san/418/console

Full run : https://jenkins.pwx.dev.purestorage.com/job/Users/job/santoshn/job/gke-pipleine-san/419/consoleFull

**Special notes for your reviewer**:
https://jenkins.pwx.dev.purestorage.com/job/Users/job/santoshn/job/gke-pipleine-san/392/

### CODE REVIEW

File: deployments/deploy-ssh.sh [hunk 1]
```diff
@@ -346,6 +346,7 @@ if [ -n "${INTERNAL_DOCKER_REGISTRY}" ]; then
 fi
 
 kubectl create configmap cloud-config --from-file=/config/cloud-json
+kubectl create configmap cloud-source-config --from-file=/cluster/cloud-source-config
```
Comment (mkoppal-px) [inline deployments/deploy-ssh.sh:349]: Why do we need this?

File: drivers/scheduler/scheduler.go [hunk 2]
```diff
@@ -347,6 +347,9 @@ type Driver interface {
 	// SetConfig sets connnection config (e.g. kubeconfig in case of k8s) for scheduler driver
 	SetConfig(configPath string) error
 
+	// SetGkeConfig sets connnection config (e.g. kubeconfig in case of k8s) for scheduler driver
```
Comment (mkoppal-px) [inline drivers/scheduler/scheduler.go:350]: Change the doc string to indicate that this is for GKE clusters

File: drivers/scheduler/k8s/k8s.go [hunk 3]
```diff
@@ -340,6 +342,90 @@ func (k *K8s) AddNewNode(newNode corev1.Node) error {
...
+		if err != nil {
+			return err
+		}
+	}
+
+	config, err := os.ReadFile(kubeconfigPath)
+	if err != nil {
+		return err
+	}
+	//configFile := base64.StdEncoding.EncodeToString(bdat)
```
Comment (mkoppal-px) [inline drivers/scheduler/k8s/k8s.go:363]: Remove the comment if not needed

File: drivers/scheduler/k8s/k8s.go [hunk 4]
```diff
@@ -340,6 +342,90 @@ func (k *K8s) AddNewNode(newNode corev1.Node) error {
...
+			Type: backupapi.CloudCredentialInfo_Google,
+			Config: &backupapi.CloudCredentialInfo_GoogleConfig{
+				GoogleConfig: &backupapi.GoogleConfig{
+					ProjectId: "portworx-eng",
+					JsonKey:   gkeJsonKey,
+				},
+			},
+		},
+	}
+	println("gke10")
```
Comment (mkoppal-px) [inline drivers/scheduler/k8s/k8s.go:403]: Remove this print

Comment (mkoppal-px) [inline drivers/scheduler/k8s/k8s.go:403]: I am not sure why are we even doing this here

File: drivers/scheduler/k8s/k8s.go [hunk 5]
```diff
@@ -340,6 +342,90 @@ func (k *K8s) AddNewNode(newNode corev1.Node) error {
...
+		CloudCredentialInfo: &backupapi.CloudCredentialInfo{
+			Type: backupapi.CloudCredentialInfo_Google,
+			Config: &backupapi.CloudCredentialInfo_GoogleConfig{
+				GoogleConfig: &backupapi.GoogleConfig{
+					ProjectId: "portworx-eng",
+					JsonKey:   gkeJsonKey,
+				},
+			},
+		},
+	}
```
Comment (mkoppal-px) [inline drivers/scheduler/k8s/k8s.go:402]: We should not do any backup operations here. This is a common function. Why do we need this?

File: tests/backup/backup_basic_test.go [hunk 6]
```diff
@@ -159,7 +160,7 @@ var _ = BeforeSuite(func() {
 			CreateBucket(provider, globalAzureBucketName)
 			log.Infof("Bucket created with name - %s", globalAzureBucketName)
 		case drivers.ProviderGke:
-			globalGCPBucketName = fmt.Sprintf("%s-%s", globalGCPBucketPrefix, bucketNameSuffix)
+			globalGCPBucketName = fmt.Sprintf("%s", globalGCPBucketPrefix)
```
Comment (mkoppal-px) [inline tests/backup/backup_basic_test.go:163]: Why was this changed?

File: tests/common.go [hunk 7]
```diff
@@ -2822,6 +2845,9 @@ func SetClusterContext(clusterConfigPath string) error {
 		if err != nil {
 			return fmt.Errorf("failed to switch to context. RefreshDriver (Node) Error: [%v]", err)
 		}
+	} else {
+		err = Inst().N.Init(node.InitOptions{
+			SpecDir: Inst().SpecDir})
```
Comment (mkoppal-px) [inline tests/common.go:2850]: Why was this needed?

File: tests/common.go [hunk 8]
```diff
@@ -8853,3 +8980,9 @@ func DeleteCrAndRepo(appData *asyncdr.AppData, appPath string) error {
 	asyncdr.DeleteCRAndUninstallCRD(appData.OperatorName, appPath, appData.Ns)
 	return nil
 }
+
+// Set default provider as aws
+func getClusterProvider() string {
+	clusterProvider = os.Getenv("CLUSTER_PROVIDER")
+	return clusterProvider
+}
```
Comment (mkoppal-px) [inline tests/common.go:8988]: Not required as we are passing this as environment variable

File: deployments/deploy-ssh.sh (hunk 1 above)
Comment (sn-px) [reply deployments/deploy-ssh.sh:349]: THis is for the spawn .. it uses different kubeconfig

File: tests/common.go (hunk 7 above)
Comment (sn-px) [reply tests/common.go:2850]: This is needed in case we are running on cloud to spin up debug pods

File: tests/common.go (hunk 8 above)
Comment (sn-px) [reply tests/common.go:8988]: Added this to make it more readable

File: drivers/scheduler/k8s/k8s.go (hunk 5 above)
Comment (sn-px) [reply drivers/scheduler/k8s/k8s.go:402]: kubeauth.UpdateClientByCredObject(&cloudCred, clientConfig, &rawConfig) needs CloudCredentialObject which is avilable in px-backup-api

File: tests/backup/backup_basic_test.go (hunk 6 above)
Comment (sn-px) [reply tests/backup/backup_basic_test.go]: fixed the code if the bucket name is empty it was generating the name with -- which is invalid

File: drivers/scheduler/k8s/k8s.go (hunk 4 above)
Comment (sn-px) [reply drivers/scheduler/k8s/k8s.go]: done

File: drivers/scheduler/scheduler.go (hunk 2 above)
Comment (sn-px) [reply drivers/scheduler/scheduler.go]: done

File: tests/common.go (hunk 7 above)
Comment (sn-px) [reply tests/common.go]: done ,added the comments

File: drivers/scheduler/k8s/k8s.go (hunk 5 above)
Comment (mkoppal-px) [reply drivers/scheduler/k8s/k8s.go:402]: We need to create our own function which will pick the values needed without needing the CloudCredentialInfo

Comment (mkoppal-px) [reply drivers/scheduler/k8s/k8s.go:402]: How will this be useful for non px-backup tests? Why will they create `CloudCredentialObject`

Comment (sn-px) [reply drivers/scheduler/k8s/k8s.go]: Done

File: drivers/scheduler/k8s/k8s.go [hunk 9]
```diff
@@ -340,6 +340,67 @@ func (k *K8s) AddNewNode(newNode corev1.Node) error {
 	return nil
 }
 
+func (k *K8s) SetGkeConfig(kubeconfigPath string, gkeCredString string) error {
```
Comment (kphalgun-px) [inline drivers/scheduler/k8s/k8s.go:343]: Is it possible to call `func (k *K8s) SetConfig(kubeconfigPath string) error` inside this?

File: tests/backup/backup_basic_test.go [hunk 10]
```diff
@@ -123,6 +124,10 @@ func BackupInitInstance() {
 	kubeconfigList := strings.Split(kubeconfigs, ",")
 	dash.VerifyFatal(len(kubeconfigList) < 2, false, "minimum 2 kubeconfigs are required for source and destination cluster")
 	DumpKubeconfigs(kubeconfigList)
+	GloabalGkeSecretString, err = GetGkeSecret()
+	if err != nil {
+		log.FailOnError(err, "Failed to fetch the gke secret")
+	}
```
Comment (kphalgun-px) [inline tests/backup/backup_basic_test.go:130]: How about using log.FailOnError directly without if condition?

**Review by kphalgun-px — APPROVED** [review]
LGTM

File: drivers/scheduler/k8s/k8s.go (hunk 9 above)
Comment (sn-px) [reply drivers/scheduler/k8s/k8s.go:343]: yes

File: tests/backup/backup_basic_test.go (hunk 10 above)
Comment (sn-px) [reply tests/backup/backup_basic_test.go]: done

File: deployments/deploy-pxb-cloud.sh [hunk 11]
```diff
@@ -333,6 +333,7 @@ if [ -n "${INTERNAL_DOCKER_REGISTRY}" ]; then
 fi
 
 kubectl create configmap cloud-config --from-file=/config/cloud-json
+kubectl create configmap cloud-source-config --from-file=/cluster/cloud-source-config
```
Comment (mkoppal-px) [inline deployments/deploy-pxb-cloud.sh:336]: Why are we creating this config map? Can you give a more meaningful name for this configmap?

File: deployments/deploy-ssh.sh (hunk 1 above)
Comment (mkoppal-px) [reply deployments/deploy-ssh.sh:349]: Why do we have the same thing in both `deploy-ssh.sh` as well as `deploy-pxb-cloud.sh`?
Do we need this for both on prem and cloud?

File: drivers/scheduler/scheduler.go [hunk 12]
```diff
@@ -347,6 +347,9 @@ type Driver interface {
 	// SetConfig sets connnection config (e.g. kubeconfig in case of k8s) for scheduler driver
 	SetConfig(configPath string) error
 
+	// SetGkeConfig sets connnection config (e.g. kubeconfig in case of GKE cluster) for scheduler driver
```
Comment (mkoppal-px) [inline drivers/scheduler/scheduler.go:350]: nit: spell check `connection`

File: drivers/scheduler/scheduler.go [hunk 13]
```diff
@@ -347,6 +347,9 @@ type Driver interface {
...
+	SetGkeConfig(configPath string, gkeCredString string) error
```
Comment (mkoppal-px) [inline drivers/scheduler/scheduler.go:351]: We can call this `jsonKey` instead of `gkeCredString` because it is called the same in Px Backup

File: drivers/scheduler/k8s/k8s.go [hunk 14]
```diff
@@ -340,6 +340,67 @@ func (k *K8s) AddNewNode(newNode corev1.Node) error {
...
+	rawConfig, err := client.RawConfig()
+	if err != nil {
+		return err
+	}
+
+	// Then create a default client config with the default loading rules
+	client = clientcmd.NewDefaultClientConfig(rawConfig, &clientcmd.ConfigOverrides{})
+	if err != nil {
+		return err
+	}
```
Comment (mkoppal-px) [inline drivers/scheduler/k8s/k8s.go:377]: Not sure why do we even need these lines of code. It is not being used anywhere.
All we need it the `clientConfig` and then load the `clientConfig` with `clientConfig.AuthProvider.Config["cred-json"]`

File: tests/common.go [hunk 15]
```diff
@@ -170,6 +170,7 @@ const (
 var (
 	clusterProviders       = []string{"k8s"}
 	GlobalCredentialConfig backup.BackupCloudConfig
+	GloabalGkeSecretString string
```
Comment (mkoppal-px) [inline tests/common.go:173]: nit: `Global`

File: tests/common.go [hunk 16]
```diff
@@ -2817,11 +2841,18 @@ func SetClusterContext(clusterConfigPath string) error {
...
+		// If it is an SSH driver, refresh the driver
 		err = ssh.RefreshDriver(sshNodeDriver)
 		if err != nil {
 			return fmt.Errorf("failed to switch to context. RefreshDriver (Node) Error: [%v]", err)
 		}
+	} else {
+		// If the Node driver is not of type SSH, initialize it to create Daemon set to spin up debug pods
+		err = Inst().N.Init(node.InitOptions{
+			SpecDir: Inst().SpecDir,
+		})
```
Comment (mkoppal-px) [inline tests/common.go:2855]: We need to discuss why this is needed because it works without this in IKS cluster where SSH does not work

**Comment by sn-px** [comment]
Closed
